### PR TITLE
Access Control Code Examples

### PR DESCRIPTION
**Pull request by janmiderback**
It's a pretty minimal set of examples and does not cover all aspects of rule writing. There are two engines. The first engine is a "read-only" engine and it is displayed how to write rules where an admin can do anything, and other users can only read/view.
The second engine also holds a minimal but slightly more complicated rule set where the "reload" access right belongs to a certain set of users. Then there are users that can create resources and users that can only view.

All access rules are tested using go test cases. The tests are stateful, which requires them to run in a certain order. While this is maybe not best practice I have gathered dependent sets in suites and used the `t.Run()` feature to control order of execution. As this is examples and not actual product tests, I think it is ok.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,24 @@
...
+
+## Running
+
+** Note: You need Go installed and the repo located correctly on the GO path <gopath>/src/github.com/qlik-oss/core-authorization
+
+
+Start the engines using docker-compose
+
+```bash
+docker-compose up -d
```
Comment (wennmo) [inline README.md]: ACCEPT_EULA?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,24 @@
...
+```
+
+Run the test cases using
+
+```bash
+go test -v ./access
```
Comment (wennmo) [inline README.md]: Need to run `dep ensure` first?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,24 @@
-# core-authorization
-[EXPERIMENTAL] Example of how to set up authorization in Qlik Core
+# [EXPERIMENTAL] Core Authorization
+
+This repo consists of two parts:
+1. Three Qlik Associative Engines set up with different access rules.
```
Comment (wennmo) [inline README.md]: Only see two engines in the docker-compose

File: .gitattributes [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# Set the default behavior, in case people don't have core.autocrlf set.
+* text=auto
+
+# Unix line endings in all text files.
+* text eol=lf
+
+*.qvf binary
```
Comment (wennmo) [inline .gitattributes]: Remove if there are no binaries in the repo

File: README.md [hunk 5]
```diff
@@ -1,2 +1,24 @@
-# core-authorization
-[EXPERIMENTAL] Example of how to set up authorization in Qlik Core
+# [EXPERIMENTAL] Core Authorization
+
```
Comment (wennmo) [inline README.md]: Perhaps also link back to the documentation in `core-website`?

**Review by wennmo — APPROVED** [review]
LGTM, just minor comments.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,24 @@
-# core-authorization
-[EXPERIMENTAL] Example of how to set up authorization in Qlik Core
+# [EXPERIMENTAL] Core Authorization
```
Comment (janmiderback) [inline README.md]: Skip this file for now. It is not ready yet.

File: access/access.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+		Info: &enigma.NxInfo{Id: name, Type: "sheet"},
+		ChildListDef: &enigma.ChildListDef{
+			Data: msg,
+		},
+	}
+
+	return doc.CreateObject(ctx, &props)
+}
+
+// readObject returns ...
```
Comment (janmiderback) [inline access/access.go]: Fix

File: access/access.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+func readObject(doc *enigma.Doc, name string) (*enigma.GenericObject, error) {
+	return doc.GetObject(ctx, name)
+}
+
+// reloadMoviesData ...
```
Comment (janmiderback) [inline access/access.go]: Fix

File: access/access.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+		return err
+	}
+	if !reloadOk {
+		return errors.New("reload failed")
+	}
+
+	return nil
+}
+
+// createMoviesObject ...
```
Comment (janmiderback) [inline access/access.go]: Fix

File: access/access.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+	err = doc.SaveObjects(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+func readMoviesData(obj *enigma.GenericObject) (titles []string, years []float64, err error) {
```
Comment (janmiderback) [inline access/access.go]: Comment

File: access/access_test.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
+package access
```
Comment (janmiderback) [inline access/access_test.go:6]: Provide file-level doc.

File: access/access_test.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+var state struct {
+	global *enigma.Global
+	app    *enigma.Doc
+	sheet  *enigma.GenericObject
+	movies *enigma.GenericObject
+	appID  string
+}
+
+// ...
```
Comment (janmiderback) [inline access/access_test.go]: Doc.

File: access/access_test.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+		// Run test cases for non-admin user.
+		t.Run("non-admin user should be able to open app", testCanOpenApp)
+		t.Run("non-admin user should be able to read sheet", testCanReadSheet)
+		t.Run("non-admin user should not be able to create sheet", testCannotCreateSheet)
+		t.Run("non-admin user should bot be able to create app", testCannotCreateApp)
+	}()
+}
+
+func TestAccessReloadEngine(t *testing.T) {
```
Comment (janmiderback) [inline access/access_test.go]: Doc.

File: access/access_test.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+		"allowReload": true,
+	}
+	viewClaims = jwt.MapClaims{
+		"sub":       "someViewUser",
+		"allowView": true,
+	}
+	err error
+)
+
+var state struct {
```
Comment (janmiderback) [inline access/access_test.go]: Document

File: README.md (hunk 1 above)
Comment (janmiderback) [reply README.md]: :+1: This file is not done.

File: access/access.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+		"meta": "/meta",
+		"order": "/order",
+		"type": "/qInfo/qType",
+		"id": "/qInfo/qId",
+		"lb": "/qListObjectDef",
+		"hc": "/qHyperCubeDef"
+	}`)
+
+	props := enigma.GenericObjectProperties{
+		Info: &enigma.NxInfo{Id: name, Type: "sheet"},
```
Comment (peol) [inline access/access.go]: I'd prefer if we keep the example generic, avoiding examples of how to do access rules on Sense-specific objects (type 'sheet', which has an 'order' property)

**Review by peol — APPROVED** [review]
LGTM, minor comments. Also (maybe I missed it), but it's quite hard to understand the golang logic (at a glance) to give them context, perhaps explain it a bit more in detail?

**Pull request merged**